libqtile/backend/wayland/core.py

```diff
@@ -0,0 +1,339 @@
+            return
+
+        wid = 0
+        wids = [win.wid for win in self.windows]
+        while True:
+            if wid not in wids:
+                break
+        win = window.Window(self, self.qtile, surface, wid)
+        logger.info(f"Managing new top-level window with window ID: {wid}")
+        self.windows.append(win)
+        self.qtile.manage(win)
+
+    def _on_cursor_axis(self, _listener, event: pointer.PointerEventAxis):
+        logger.debug("Signal: cursor axis")
+        self.seat.pointer_notify_axis(
+            event.time_msec, event.orientation, event.delta, event.delta_discrete, event.source,
+        )
+
+    def _on_cursor_frame(self, _listener, _data):
+        logger.debug("Signal: cursor frame")
+        self.seat.pointer_notify_frame()
+
+    def _on_cursor_button(self, _listener, event: pointer.PointerEventButton):
+        logger.debug("Signal: cursor button")
+        self.seat.pointer_notify_button(
+            event.time_msec, event.button, event.button_state
+        )
+
+    def _on_cursor_motion(self, _listener, event: pointer.PointerEventMotion):
+        assert self.qtile is not None
+        logger.debug("Signal: cursor motion")
+        self.cursor.move(event.delta_x, event.delta_y, input_device=event.device)
+        self.qtile.process_button_motion(self.cursor.x, self.cursor.y)
+
+    def _on_cursor_motion_absolute(self, _listener, event: pointer.PointerEventMotionAbsolute):
+        assert self.qtile is not None
+        logger.debug("Signal: cursor motion_absolute")
+        self.cursor.warp(
+            WarpMode.AbsoluteClosest, event.x, event.y, input_device=event.device,
+        )
+        self.qtile.process_button_motion(self.cursor.x, self.cursor.y)
+
+    def _add_new_pointer(self, device: input_device.InputDevice):
+        logger.info("Adding new pointer")
+        self.cursor.attach_input_device(device)
+
+    def _add_new_keyboard(self, device: input_device.InputDevice):
+        logger.info("Adding new keyboard")
+        self.keyboards.append(keyboard.Keyboard(self, device))
+        self.seat.set_keyboard(device)
+
+    def focus_window(self, win: window.Window, surface: Surface = None):
+        if surface is None:
+            surface = win.surface.surface
+
+        previous_surface = self.seat.keyboard_state.focused_surface
+        if previous_surface == surface:
+            return
+
+        if previous_surface is not None:
+            # Deactivate the previously focused surface
+            previous_xdg_surface = xdg_shell.XdgSurface.from_surface(previous_surface)
+            previous_xdg_surface.set_activated(False)
+
+        # roll the given surface to the front of the list, copy and modify the
+        # list, then save back to prevent any race conditions on list
+        # modification
+        windows = self.windows[:]
+        windows.remove(win)
+        windows.append(win)
+        self.win = windows
```

Comment:
I think a lot of this logic in the tiny compositor is just to facilitate being able to do simple focus prev/focus next like operations. I would think that would be handled somewhere else once we are in qtile. Does that sound right to you?

Reply:
Yep I think so too. The Core is maintaining a list of windows that is used only to know what order to render them and currently has no awareness of groups or anything like that. The way it manages windows will have to change (whether this is per group or not is too early to tell). I have a working setup for "virtual desktops" on another project (https://github.com/m-col/wimp) which I'll try to mirror as the PR as outgrown tinywm.